Fix compatibility issues + bugs

lorenzhawkes opened the pull request:
1) From LSP spec [3.16 to 3.17](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#workspace_symbol), dropping the location attribute from the
   response to `workspace/symbol` is only permitted if the client
   capability `workspace.symbol.resolveSupport` is advertized. Without
   this setting, the `location` attribute should be present.
2) `findDeclarationCursorSite` constructs a location using a non-uri
   path in the uri property. This leads to document symbol rename
   for example to not work.
3) THe stdio backend for logging accidentally overrides the stdout
   backend with the one meant for stderr.

Commit:
Fix compatibility issues + bugs

1) From LSP spec 3.16 to 3.17, dropping the location attribute from the
   response to `workspace/symbol` is only permitted if the client
   capability `workspace.symbol.resolveSupport` is advertized. Without
   this setting, the `location` attribute should be present.
2) `findDeclarationCursorSite` constructs a location using a non-uri
   path in the uri property. This leads to document symbol rename
   for example to not work.
3) THe stdio backend for logging accidentally overrides the stdout
   backend with the one meant for stderr.

Commit:
Fallback to declaration text range

Not all named declarations have a name identifier. In those cases
fallback to using the range that encloses the whole named declaration
rather than returning nothing.

Commit:
Fix detekt violations

fwcd commented on server/src/main/kotlin/org/javacs/kt/KotlinLanguageServer.kt:
```diff
@@ -141,6 +143,11 @@ class KotlinLanguageServer(
         InitializeResult(serverCapabilities, serverInfo)
     }
 
+    private fun clientHasWorkspaceSymbolResolveSupport(clientCapabilities: ClientCapabilities) =
+        clientCapabilities?.workspace?.symbol?.resolveSupport?.properties?.let { properties ->
+            if (properties.size > 0) SymbolResolveSupport(true, properties) else null
+        } ?: SymbolResolveSupport(false, emptyList())
```
Nit: If the function does not return a boolean, its name should not be `clientHas...`. Though if we pass the client capabilities to `KotlinWorkspaceService`, I would either

- remove this function (since we only use it in one place anyway)
- move `SymbolResolveSupport` to a separate file (e.g. in `org.javacs.kt.symbols`) and make this an extension property on `ClientCapabilities`, i.e. something along the lines of
  ```kotlin
  val ClientCapabilities.symbolResolveSupport
    get() = ...
  ```

lorenzhawkes replied:
I went for the SymbolResolveSupport. I think technically more properties can me masked in the symbol calls in the future although the only example of this is the location property, based on the types available for return.

fwcd commented on server/src/main/kotlin/org/javacs/kt/symbols/Symbols.kt:
```diff
@@ -33,10 +34,10 @@ private fun doDocumentSymbols(element: PsiElement): List<DocumentSymbol> {
 }
 
-fun workspaceSymbols(query: String, sp: SourcePath): List<WorkspaceSymbol> =
+fun workspaceSymbols(locationRequired: Boolean, query: String, sp: SourcePath): List<WorkspaceSymbol> =
```
Style nit: I would pass the `locationRequired` parameter last, to maintain the ordering of decreasing significance.

fwcd commented on server/src/main/kotlin/org/javacs/kt/symbols/Symbols.kt:
```diff
@@ -56,10 +57,28 @@ private fun pickImportantElements(node: PsiElement, includeLocals: Boolean): KtN
 
-private fun workspaceSymbol(d: KtNamedDeclaration): WorkspaceSymbol? {
-    val name = d.name ?: return null
+private fun workspaceSymbol(locationRequired: Boolean): (KtNamedDeclaration) -> WorkspaceSymbol? {
```
Another style nit: Most functions in this codebase are not curried, therefore I would prefer the uncurried versions here too (even though it requires using a lambda at the call site):

```suggestion
private fun workspaceSymbol(declaration: KtNamedDeclaration, locationRequired: Boolean): WorkspaceSymbol? {
```

This would keep the naming pattern more consistent too (most singular thing-named functions also return the thing in question and not a function).

fwcd reviewed (commented):
Thanks, I have left some comments below

Commit:
Requested changes

1) pass client capabilities into KotlinWorkspaceService
2) Moved SymbolResolveSupport to a new file and added extension property
   to ClientCapabilities for it
3) Added a test for the location fetching required by the symbol resolve
   client capability

lorenzhawkes commented on server/src/main/kotlin/org/javacs/kt/KotlinLanguageServer.kt:
```diff
@@ -108,6 +108,8 @@ class KotlinLanguageServer(
             serverCapabilities.renameProvider = Either.forRight(RenameOptions(false))
         }
 
+        config.workspace.symbolResolveSupport = clientHasWorkspaceSymbolResolveSupport(clientCapabilities)
```
no problem, I change it to pass the client capabilities into the KotlinWorkspaceService
